clang/lib/AST/PreorderAST.cpp

```diff
@@ -36,56 +36,57 @@ void PreorderAST::AddNode(Node *N, Node *Parent) {
     I->Child = N;
 }
 
-bool PreorderAST::CanCoalesceNode(OperatorNode *O) {
-  if (!O || !isa<OperatorNode>(O) || !O->Parent)
+bool PreorderAST::CanCoalesceNode(BinaryOperatorNode *B) {
+  if (!B || !isa<BinaryOperatorNode>(B) || !B->Parent)
     return false;
 
   // We can only coalesce if the operator of the current and parent node is
   // commutative and associative. This is because after coalescing we later
   // need to sort the nodes and if the operator is not commutative and
   // associative then sorting would be incorrect.
-  if (!O->IsOpCommutativeAndAssociative())
+  if (!B->IsOpCommutativeAndAssociative())
     return false;
-  auto *OParent = dyn_cast_or_null<OperatorNode>(O->Parent);
-  if (!OParent || !OParent->IsOpCommutativeAndAssociative())
+  auto *BParent = dyn_cast_or_null<BinaryOperatorNode>(B->Parent);
+  if (!BParent || !BParent->IsOpCommutativeAndAssociative())
     return false;
 
   // We can coalesce in the following scenarios:
   // 1. The current and parent nodes have the same operator OR
   // 2. The current node is the only child of its operator node (maybe as a
   // result of constant folding).
-  return O->Opc == OParent->Opc || O->Children.size() == 1;
+  return B->Opc == BParent->Opc || B->Children.size() == 1;
 }
 
-void PreorderAST::CoalesceNode(OperatorNode *O) {
-  if (!CanCoalesceNode(O)) {
+void PreorderAST::CoalesceNode(BinaryOperatorNode *B) {
+  if (!CanCoalesceNode(B)) {
     assert(0 && "Attempting to coalesce invalid node");
     SetError();
     return;
   }
 
-  // If the current node can be coalesced, its parent must be an OperatorNode.
-  auto *OParent = dyn_cast_or_null<OperatorNode>(O->Parent);
-  if (!OParent)
+  // If the current node can be coalesced, its parent must be a
+  // BinaryOperatorNode.
+  auto *BParent = dyn_cast_or_null<BinaryOperatorNode>(B->Parent);
+  if (!BParent)
     return;
 
   // Remove the current node from the list of children of its parent.
-  for (auto I = OParent->Children.begin(),
-            E = OParent->Children.end(); I != E; ++I) {
-    if (*I == O) {
-      OParent->Children.erase(I);
+  for (auto I = BParent->Children.begin(),
```

Comment:
Because `BParent` is mutated in the loop we need to compute its end each time through the loop. Currently we are not doing so because we compute `E = BParent->Children.end()` in the loop initializer. So the correct initializer should be as follows:
`for (auto I = BParent->Children.begin(); I != BParent->Children.end(); ++I)`

Could you also add a comment here indicating that the reason we need to evaluate loop end each time is because the container is mutated in the loop?

For more details see https://llvm.org/docs/CodingStandards.html#don-t-evaluate-end-every-time-through-a-loop